### ISSUE TITLE
Task restarts though set to stop on fail

### ISSUE THREAD

**Issue by tommyalatalo** [hashicorp/nomad#12017]
### Nomad version
Nomad v1.2.4 (9f21b724acf115a58d9ecdc97d281aef01d36387)

### Issue

Nomad keeps restarting tasks that are set to stop on failure and not retry further.
My configuration includes this according to the docs [disabling rescheduling](https://www.nomadproject.io/docs/job-specification/reschedule#disabling-rescheduling) and [restart mode](https://www.nomadproject.io/docs/job-specification/restart#mode-values)

```
    reschedule {
      attempts  = 0
      unlimited = false
    }

    restart {
      mode = "fail"
    }
```

### Reproduction steps
Set up any failing task with the above restart/reschedule configuration.

#### Expected Result
The job should fail on the first try and not try to reschedule or restart the task.
 
#### Actual Result
Nomad restarts the task several times.

**Comment by tgross** [comment]
Hi @tommyalatalo! That example for [disabling rescheduling](https://www.nomadproject.io/docs/job-specification/reschedule#disabling-rescheduling) disables _rescheduling_, which is when the scheduler tries to place a new allocation for the failed allocation. 

But the configuration you have for `restart` doesn't prevent _restarts_, which is when the client restarts a task in place. The docs could probably be a little more clear on this, but if you look at these two sections together:

> [`mode`](https://www.nomadproject.io/docs/job-specification/restart#mode) `(string: "fail")` Controls the behavior when the task fails more than attempts times in an interval. For a detailed explanation of these values and their behavior, please see the [mode values section](https://www.nomadproject.io/docs/job-specification/restart#mode-values).
> ...
> ["fail"](https://www.nomadproject.io/docs/job-specification/restart#fail) - Instructs the scheduler to not attempt to restart the task on failure. This is the default behavior. This mode is useful for non-idempotent jobs which are unlikely to succeed after a few failures. Failed jobs will be restarted according to the [reschedule](https://www.nomadproject.io/docs/job-specification/reschedule) stanza.

The config you've shown is missing an `attempts` field, which means it's getting the default behavior (which [varies depending on the job type](https://www.nomadproject.io/docs/job-specification/restart#restart-parameter-defaults)).

Try this:

```hcl
reschedule {
  attempts  = 0
  unlimited = false
}

restart {
  attempts = 0
  mode = "fail"
}
```

If that works for you, it'd probably be a good idea for us to describe this a little better in the docs. I'll mark this as a docs issue to clean up.

**Comment by tommyalatalo** [comment]
> Hi @tommyalatalo! That example for [disabling rescheduling](https://www.nomadproject.io/docs/job-specification/reschedule#disabling-rescheduling) disables _rescheduling_, which is when the scheduler tries to place a new allocation for the failed allocation.
> 
> But the configuration you have for `restart` doesn't prevent _restarts_, which is when the client restarts a task in place. The docs could probably be a little more clear on this, but if you look at these two sections together:
> 
> > [`mode`](https://www.nomadproject.io/docs/job-specification/restart#mode) `(string: "fail")` Controls the behavior when the task fails more than attempts times in an interval. For a detailed explanation of these values and their behavior, please see the [mode values section](https://www.nomadproject.io/docs/job-specification/restart#mode-values).
> > ...
> > ["fail"](https://www.nomadproject.io/docs/job-specification/restart#fail) - Instructs the scheduler to not attempt to restart the task on failure. This is the default behavior. This mode is useful for non-idempotent jobs which are unlikely to succeed after a few failures. Failed jobs will be restarted according to the [reschedule](https://www.nomadproject.io/docs/job-specification/reschedule) stanza.
> 
> The config you've shown is missing an `attempts` field, which means it's getting the default behavior (which [varies depending on the job type](https://www.nomadproject.io/docs/job-specification/restart#restart-parameter-defaults)).
> 
> Try this:
> 
> ```terraform
> reschedule {
>   attempts  = 0
>   unlimited = false
> }
> 
> restart {
>   attempts = 0
>   mode = "fail"
> }
> ```
> 
> If that works for you, it'd probably be a good idea for us to describe this a little better in the docs. I'll mark this as a docs issue to clean up.

Yes, I've read the docs several times on different occasions and this is what the failure mode description is for "fail":

_"fail" - **Instructs the scheduler to not attempt to restart the task on failure**. This is the default behavior. This mode is useful for non-idempotent jobs which are unlikely to succeed after a few failures. Failed jobs will be restarted according to the [reschedule](https://www.nomadproject.io/docs/job-specification/reschedule) stanza._

The above is the description in its entirety. Since the corresponding example also doesn't show `attempts = 0` one can only assume that setting the mode to `fail` should stop the task from restarting when you read the description above. This most definitely needs clarification as the text/example now it missing a key component.

**Comment by dvusboy** [comment]
The key phrase in the docs is under the [`restart` Parameters](https://www.nomadproject.io/docs/job-specification/restart#restart-parameters) section. In there it states: `mode` *controls the behavior when the task fails more than `attempts` times in an interval*. That drives the behavior of the `restart` parameters. I can see how the explanation on `mode` itself may confuse.

**Comment by tgross** [comment]
Yeah it's there... but not clear. Having `delay` mean "just wait for the next interval" and not specifically calling out that it doesn't get rescheduled is confusing. Splitting the language for the mode behavior between the `mode` field description and the `fail` setting is confusing. I'll rework this bit of docs and open a PR.

**Comment by tgross** [comment]
Closed by https://github.com/hashicorp/nomad/pull/12275
